### PR TITLE
FYST-1209-add-voluntary-donations-screen-and-xml-pdf-functionality

### PR DESCRIPTION
**Pull request by DrewProebstel**
## Link to pivotal/JIRA issue
https://codeforamerica.atlassian.net/browse/FYST-1209## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!

**Reminder**: merge main into this branch and get green tests before merging to main
## What was done?
-  Adds the Id Donations Page
- Adds navigation from Id Review to this page
## How to test?
- Heroku
- RSpec

## Screenshots (for visual changes)
Screenshot 2024-11-27 at 3 10 19 PM

Screenshot 2024-11-27 at 3 13 35 PM

Screenshot 2024-11-27 at 3 13 07 PM

### CODE REVIEW

File: app/controllers/state_file/questions/id_donations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+module StateFile
+  module Questions
+    class IdDonationsController < QuestionsController
+      def self.show?(intake)
+        false
```
Comment (powersurge360) [inline app/controllers/state_file/questions/id_donations_controller.rb:5]: Always false?

File: app/controllers/state_file/questions/id_donations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+      end
+
+      private
+
+      def next_path
+        StateFile::Questions::IdReviewController.to_path_helper
```
Comment (powersurge360) [inline app/controllers/state_file/questions/id_donations_controller.rb:11]: Why override this here vs registering in the navigation class?

File: app/lib/navigation/state_file_id_question_navigation.rb [hunk 3]
```diff
@@ -38,6 +38,7 @@ class StateFileIdQuestionNavigation < Navigation::StateFileBaseQuestionNavigatio
         Navigation::NavigationStep.new(StateFile::Questions::PrimaryStateIdController),
         Navigation::NavigationStep.new(StateFile::Questions::SpouseStateIdController),
         Navigation::NavigationStep.new(StateFile::Questions::IdReviewController),
+        Navigation::NavigationStep.new(StateFile::Questions::IdDonationsController, false), # show? false
```
Comment (powersurge360) [inline app/lib/navigation/state_file_id_question_navigation.rb:41]: Why skip show?

File: app/forms/state_file/id_donations_form.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    )
+
+    validates :nongame_wildlife_fund_donation, numericality: { greater_than_or_equal_to: 0 }, allow_blank: true
+    validates :childrens_trust_fund_donation, numericality: { greater_than_or_equal_to: 0 }, allow_blank: true
+    validates :special_olympics_donation, numericality: { greater_than_or_equal_to: 0 }, allow_blank: true
+    validates :guard_reserve_family_donation, numericality: { greater_than_or_equal_to: 0 }, allow_blank: true
+    validates :american_red_cross_fund_donation, numericality: { greater_than_or_equal_to: 0 }, allow_blank: true
+    validates :veterans_support_fund_donation, numericality: { greater_than_or_equal_to: 0 }, allow_blank: true
+    validates :food_bank_fund_donation, numericality: { greater_than_or_equal_to: 0 }, allow_blank: true
+    validates :opportunity_scholarship_program_donation, numericality: { greater_than_or_equal_to: 0 }, allow_blank: true
```
Comment (powersurge360) [inline app/forms/state_file/id_donations_form.rb:22]: [dust] You may consider consolidating these validations. They're all the same, to my eye, so they could be specified on a single line or, if you prefer readability, you may consider capturing them all into an array and then calling something like

```ruby
validates *non_zero_required_fields, numericality: { greater_than_or_equal_to: 0 }, allow_blank: true
```

It's a minor matter of taste though tbh

**Review by powersurge360 — COMMENTED** [review]
I am unsure about the skipping of the controller always and the explicit registration of a next path. Would like to make sure that's on purpose and what the use case is. Also included a minor suggestion for validations (which is of the weakest suggestion variety). Totally fine leaving it as is.

**Review by powersurge360 — APPROVED** [review]
Drew reached out and let me know that this is a weird pocket dimension screen that gets skipped in the standard flow and can only be accessed explicitly. Approving.

**Review by mrotondo — APPROVED** [review]
looks good, couple minor style things around nil/zero checking

**Pull request merged**